backend/layers/thirdparty/s3_provider.py

```diff
@@ -48,6 +48,12 @@ def upload_file(self, src_file: str, bucket_name: str, dst_file: str, extra_args
             ExtraArgs=extra_args,
         )
 
+    def delete_files(self, bucket_name: str, object_keys: List[str]):
+        """
+        Deletes the objects `object_keys` from bucket `bucket_name`
+        """
+        self.client.delete_objects(Bucket=bucket_name, Delete={"Objects": [{"Key": key} for key in object_keys]})
```

Comment:
Also add unit tests for the above cases.

Reply:
> We don't want to silently fail to delete. If we do fail we need to roll back the and database changes made. t's is an edge case, but how difficult are these datasets to track down if this fails silently?

In some cases I would agree with you, but there are some important considerations: the user is _trying_ to delete these assets; the public dataset deletion is occurring before the updates to entities in the database, so we guarantee that we don't accidentally leave datasets behind -- the request only fails silently if the key doesn't exist, which does not constitute a problem in this scenario. If the actual attempt to delete fails, perhaps for network reasons, the request fails loudly and the Collection will not be deleted from the db. The user sees 500. The user will therefore be aware that the Collection still exists. Perhaps we should capture 500s for deletes and inform them that public assets may be missing.

Actually, in lieu of that, we can perform a HEAD request for all asset files for all `dataset.assets[]` objects. That way we can say whether or not the public asset is still available. If it's not, then the curator can get in touch with us.

cc @nayib-jose-gloria for thoughts on this issue

Reply:
I changed my mind we should make the request without `quite` and log the response. Knowing what request deleted a file could help with debugging and audits in the future.

Reply:
I think capturing 500s and returning a message to the user that tombstoning failed, but some public assets may have already been deleted, and suggesting they try again is a reasonable route. Perhaps we should introduce our own retry (unless boto3 already has that as a default?) here to account for transient network blips

Reply:
Agree w/ logging responses for debugging should it come up

Reply:
> returning a message to the user that tombstoning failed, but some public assets may have already been deleted, and suggesting they try again is a reasonable route.

That sounds like ambiguity/complexity is being pushed to the user. The operation should either succeed or not?

Reply:
There's always the rare possibility that a network error causes some object(s) not to be deleted in a bulk delete request. We can enable dataset-assets to be a versioned bucket, which would allow us to restore deleted objects and rollback if we come across an error--but that seems like overengineering for this edge case IMO 

We can alternatively bulk delete and then alert engineers of an error if the boto3 delete_objects command fails on any asset + return the list of objects it failed to delete, like we do with upload errors. And users can see 500 and follow the same workflow as with upload errors

Reply:
>That sounds like ambiguity/complexity is being pushed to the user. The operation should either succeed or not?

That is true, and in all likelihood we will never run into this problem. Really what we're dealing with here is the unlikely event that _some_ of the assets are deleted successfully and then a transient error causes the whole operation to return 500 (meaning the Collection entity does not actually get marked as tombstoned/deleted), at which point we'd be left with the Collection still in play and some of the assets gone, but not all. We don't have a way to perfectly transactionalize this operation since it is db + s3 without putting a lot of effort into it.

Plus, the ambiguity will only ever get put on the user _if_ they have tried to delete the Collection (which means they would surely just try to delete it again anyways). I don't want us to spend time solving perfectly the scenario where someone tries to delete something, it fails, and then they decide they don't want to delete it anymore. 😄

Reply:
@Bento007 @nayib-jose-gloria  I actually think further optimization here is not worth the effort, and here's why: the curator only performs this action if they intend to erase the Collection. If the operation fails, i.e., if any of the deletes don't succeed, then the curator will be made aware of this in the UI because the request will fail. Therefore they've been made aware that the Collection still exists. They'll either try again, and this code will succeed or fail (try, try again), and then they'll contact us if it continues failing. When they contact us, our move will be to delete everything, which is extremely easy to do with a couple postgres queries for dataset version ids; it will not even be worth referencing the logs to figure out exactly what happened with the failed delete attempts. cc @brianraymor

Reply:
So I submit that we leave it with the current level of logging. What really matters is having the s3 deletes coming _before_ the tombstoning of the Collection entity, which is what this code is doing.

Reply:
Actually, Dan, are you sure it won't fail silently for all errors? Per the [boto3 docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/delete_objects.html), it seems like it returns the same object response shape regardless, which provides list of Errors. So I think we'd need to check here for `if resp['Errors']` and raise an exception if list has entries

Reply:
It fails loudly if the request is malformed, if permissions are not correct, etc.

So if the key doesn't exist, it fails silently. But that would indicate a business logic failing... and our standard for interactions with s3 so far has not been to double check the key name for, say, writes.

Reply:
I can't imagine a network failure would be swallowed by the boto client; that would surely be a loud failure as well.

Reply:
It 'fails loudly' by returning the error in the Errors list though, not by throwing an Exception. And if we don't check the Errors list for entries, aren't we failing silently as a result?

Reply:
Fail loudly == throws Exception